File: Program's_Contributed_By_Contributors/Python_Programs/bst_sort.py
"""
Case 1:
Enter the list of numbers: 4 2 3 4 5 6 1 7 9 10 3
Sorted list: 1 2 3 3 4 4 5 6 7 9 10 
 
Case 2:
Enter the list of numbers: 6 5 4 3 2 1
Sorted list: 1 2 3 4 5 6 
 
Case 3:
Enter the list of numbers: 5
Sorted list: 5
"""


class BSTNode:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None
        self.parent = None

    def insert(self, node):
        if self.key > node.key:
            if self.left is None:
                self.left = node
                node.parent = self
            else:
                self.left.insert(node)
        elif self.key <= node.key:
            if self.right is None:
                self.right = node
                node.parent = self
            else:
                self.right.insert(node)

    def inorder(self):
        if self.left is not None:
            self.left.inorder()
        print(self.key, end=' ')
        if self.right is not None:
            self.right.inorder()


class BSTree:
    def __init__(self):
        self.root = None

    def inorder(self):
        if self.root is not None:
            self.root.inorder()

    def add(self, key):
        new_node = BSTNode(key)
        if self.root is None:
            self.root = new_node
        else:
            self.root.insert(new_node)


bstree = BSTree()

alist = input('Enter the list of numbers: ').split()
alist = [int(x) for x in alist]
for x in alist:
    bstree.add(x)
print('Sorted list: ', end='')
bstree.inorder()

File: Program's_Contributed_By_Contributors/Implementation_of_DS/BSTsort.c
//sorting using binary tree
#include <stdio.h>
#include <stdlib.h>

struct node{
    struct node *left;
    int data;
    struct node *right;
};

void insert(struct node *, int element);

int * traversal(struct node *node ,int *sort, int *count);

int main()
{
    unsigned int n;
    printf("Enter elements you want to add in array: ");
    scanf("%d", &n);
    int arr[n];
    printf("Enter elements now: \n");

    for(int i=0;i<n;++i)
        scanf("%d", &arr[i]);

    printf("Element entered in arr: \n");
    for(int i=0;i<n;++i)
            printf("%d\t", arr[i]);
        printf("\n");


    struct node *root;
    root=(struct node *)malloc(sizeof(struct node));
    root->left=NULL;
    root->right=NULL;
    root->data=arr[0];


    for(int i=1;i<n;++i)
        insert(root, arr[i]);






    //Traversal
    int *sort;
    sort=malloc((sizeof(int)*n));
    int count=0;

    sort=traversal(root, sort, &count);

    for(int i=0;i<n;++i){
        printf("%d\t", *(sort+i));
    }

    return 0;
}


void insert(struct node *node, int element){
    if(element<node->data){
        if(node->left==NULL){
            struct node *newNode;
            newNode=(struct node *)malloc(sizeof(struct node));
            newNode->data=element;
            newNode->left=NULL;
            newNode->right=NULL;

            node->left=newNode;
        }
        else{
            insert(node->left, element);
        }
    }
    else if(element>=node->data){
        if(node->right==NULL){
            struct node *newNode;
            newNode=(struct node *)malloc(sizeof(struct node));
            newNode->data=element;
            newNode->left=NULL;
            newNode->right=NULL;

            node->right=newNode;
        }
        else{
            insert(node->right, element);
        }

    }
}


int * traversal(struct node *node, int *sort, int *count){
    if(node!=NULL){
        sort=traversal(node->left, sort, count);
        *(sort+ *count)=node->data;
        (*count)++;
        printf("\n");
        sort=traversal(node->right, sort, count);

    }
    return sort;

}
